crew

```diff
@@ -627,36 +627,28 @@ end
 
 def expand_dependencies
   @dependencies = []
-
   def push_dependencies
     if @pkg.is_binary?(@device[:architecture]) ||
        (!@pkg.in_upgrade && !@pkg.build_from_source && @device[:installed_packages].any? { |pkg| pkg[:name] == @pkg.name })
       # retrieve name of dependencies that doesn't contain :build tag
       check_deps = @pkg.dependencies.select {|k, v| !v.include?(:build)}.map {|k, v| k}
-    elsif @pkg.is_fake?
-      # retrieve name of all dependencies
-      check_deps = @pkg.dependencies.map {|k, v| k}
     else
       # retrieve name of all dependencies
       check_deps = @pkg.dependencies.map {|k, v| k}
     end
-
-    # remove a dependent package which is equal to the target
-    check_deps.select! {|name| @pkgName != name}
-
-    # add new dependencies at the beginning of array
-    @dependencies = check_deps.clone.concat(@dependencies)
-
     # check all dependencies recursively
     check_deps.each do |dep|
-      search dep, true
-      push_dependencies
+      # build unique dependencies list
+      unless @dependencies.include?(dep) || dep == @pkgName
+        @dependencies << dep
+        search dep, true
+        push_dependencies
+      end
     end
   end
-
   push_dependencies
-
-  @dependencies.uniq
+  # resolve_dependencies needs expand_dependencies to return @dependencies
+  return @dependencies
```

Comment:
Actually, you don't need to return the value here since @dependencies is an instance variable and is available globally.  This looks really good @dpgettings.  Great work!  The only thing this needs now is to bump the `CREW_VERSION` in `lib/const.rb` since this changes the crew command.

Reply:
I guess I'm confused about the return thing. Looking down at the next method, `resolve_dependencies`, it looks like `dependencies` here is actually a local variable being set by a return value from `expand_dependencies`. It seems like getting rid of the return in `expand_dependencies` would necessitate changing every occurrence of `dependencies` in the `resolve_dependencies` method into a reference to the instance variable `@dependencies`, i.e.: 
```ruby
def resolve_dependencies
  #dependencies = expand_dependencies
  expand_dependencies  # calling without an assignment
  # leave only not installed packages in dependencies
  @dependencies.select! {|name| @device[:installed_packages].none? {|pkg| pkg[:name] == name}}
  return if @dependencies.empty?
  puts "The following packages also need to be installed: "
  @dependencies.each do |dep|
    print dep + " "
  end
  puts ""
  print "Do you agree? [Y/n] "
  response = STDIN.getc
  case response
  when "n"
    abort "No changes made."
  when "\n", "y", "Y"
    puts "Proceeding..."
    proceed = true
  else
    puts "I don't understand `#{response}`. :(".lightred
    abort "No changes made."
  end
  if proceed
    @dependencies.each do |dep|
      search dep
      install
    end
  end
end
def install
  if !@pkg.in_upgrade && @device[:installed_packages].any? { |pkg| pkg[:name] == @pkg.name }
    puts "Package #{@pkg.name} already installed, skipping...".lightgreen
    return
  end
```
... Otherwise, `resolve_dependencies`'s local variable `dependencies` would be set to whatever the `expand_dependencies` method happened to implicitly return (if anything), such as the return value of the last call to `push_dependencies`.

Am I totally off base here? I'm definitely not experienced enough in Ruby to be totally sure I know what I'm talking about, haha.

Reply:
@uberhacker If you want to type something without accidentally pinging a user, surround it in backticks, e.g. `@dependencies`.

Reply:
@dpgettings: Does it work without setting the local variable?

Reply:
@dpgettings: I tested the code you have there and it works great.  I can merge this if you push those changes along with bumping `CREW_VERSION` in `lib/const.rb`.

Reply:
@dpgettings?

Reply:
@uberhacker Can we just merge this? I don't think it would require a version bump, as it doesn't remove any dependencies.

Reply:
It changes the crew command which is what drives the need to increment the version.

Reply:
All right, at least the patch version. Could you do it? It doesn't seem like he's responding.

Reply:
@skycocker